File: server/src/main/java/server/api/tls/NonBlockingSecuredConnectionHandler.java
package server.api.tls;

import javax.net.ssl.SSLSession;
import javax.net.ssl.SSLSocket;
import java.io.*;

/**
 * Created by prayagupd
 * on 11/14/16.
 */

// Thread handling the socket from client
public class NonBlockingSecuredConnectionHandler extends Thread {

    private SSLSocket securedSocket = null;

    public NonBlockingSecuredConnectionHandler(SSLSocket securedSocket) {
        this.securedSocket = securedSocket;
    }

    public void run() {
        securedSocket.setEnabledCipherSuites(securedSocket.getSupportedCipherSuites());

        try {
            // Start handshake
            securedSocket.startHandshake();

            // Get session after the connection is established
            SSLSession tlsSession = securedSocket.getSession();

            System.out.println("[INFO] NonBlockingSecuredConnectionHandler TLSSession :");
            System.out.println("\tProtocol : " + tlsSession.getProtocol());
            System.out.println("\tCipher suite : " + tlsSession.getCipherSuite());

            // Start handling application content
            InputStream receivingStream = securedSocket.getInputStream();
            OutputStream sendStream = securedSocket.getOutputStream();

            BufferedReader receivingBuffer = new BufferedReader(new InputStreamReader(receivingStream));
            PrintWriter response = new PrintWriter(new OutputStreamWriter(sendStream));

            String line = null;
            while ((line = receivingBuffer.readLine()) != null) {
                if (line.trim().isEmpty()) {
                    break;
                }
                System.out.println("[INFO] NonBlockingSecuredConnectionHandler Server consumes : " + line);
            }

            // Write data to sending response
            response.print("HTTP/1.1 200\r\n");
            response.flush();

            securedSocket.close();
        } catch (Exception ex) {
            ex.printStackTrace();
            PrintWriter printWriter = new PrintWriter(new StringWriter());
            printWriter.println("HTTP/1.1 401\r\n");
            printWriter.flush();
        }
    }
}

File: server/src/main/kotlin/server/api/tls/NonBlockingSecuredConnectionHandler.kt
package server.api.tls

import java.io.*
import javax.net.ssl.SSLSocket

/**
 * Created by prayagupd
 * on 11/14/16.
 */

// Thread handling the socket from client
class NonBlockingSecuredConnectionHandler(var securedSocket: SSLSocket) : Thread() {

    override fun run() {
        val supportedCipherSuites = securedSocket.supportedCipherSuites

        securedSocket.enabledCipherSuites = supportedCipherSuites

        try {
            // Start handshake
            securedSocket.startHandshake()

            // Get session after the connection is established
            val tlsSession = securedSocket.session

            println("[INFO] NonBlockingSecuredConnectionHandler TLSSession :")
            println("\tProtocol : " + tlsSession?.protocol)
            println("\tCipher suite : " + tlsSession?.cipherSuite)

            // Start handling application content
            val receivingStream = securedSocket.inputStream
            val sendStream = securedSocket.outputStream

            val receivingBuffer = BufferedReader(InputStreamReader(receivingStream))
            val response = PrintWriter(OutputStreamWriter(sendStream))

            var line = receivingBuffer.readLine()
            while (line!= null) {
                if (line.trim { it <= ' ' }.isEmpty()) {
                    break
                }
                println("[INFO] NonBlockingSecuredConnectionHandler Server consumes : " + line)
                line = receivingBuffer.readLine()
            }

            // Write data to sending response
            response.print("HTTP/1.1 200\r\n")
            response.flush()

            securedSocket.close()
        } catch (ex: Exception) {
            ex.printStackTrace()
            val printWriter = PrintWriter(StringWriter())
            printWriter.println("HTTP/1.1 401\r\n")
            printWriter.flush()
        }

    }
}
